packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -0,0 +1,254 @@
+}
+
+interface ComboBoxStateProps<T> extends ComboBoxProps<T> {
+  collator: Intl.Collator
+}
+
+function filter<T>(nodes: Iterable<Node<T>>, filterFn: (node: Node<T>) => boolean): Iterable<Node<T>> {
+  let filteredNode = [];
+  for (let node of nodes) {
+    if (node.type === 'section' && node.hasChildNodes) {
+      let copyOfNode = {...node};
+      let copyOfChildNodes = copyOfNode.childNodes;
+      let filtered = filter(copyOfChildNodes, filterFn);
+      if ([...filtered].length > 0) {
+        copyOfNode.childNodes = filtered;
+        filteredNode.push(copyOfNode);
+      }
+    } else if (node.type !== 'section' && filterFn(node)) {
+      filteredNode.push(node);
+    }
+  }
+  return filteredNode;
+}
+
+export function useComboBoxState<T extends object>(props: ComboBoxStateProps<T>): ComboBoxState<T> {
+  let {
+    onFilter,
+    collator,
+    onSelectionChange
+  } = props;
+
+  let [isFocused, setFocused] = useState(false);
+  let itemsControlled = !!onFilter;
+
+  let computeKeyFromValue = (value, collection) => {
+    let key;
+    for (let [itemKey, node] of collection.keyMap) {
+      if (node.type !== 'section') {
+        let itemText = node.textValue;
+        if (itemText === value) {
+          key = itemKey;
+          break;
+        }
+      }
+    }
+
+    return key;
+  };
+
+  let builder = useMemo(() => new CollectionBuilder<T>(), []);
+  let collection = useMemo(() => {
+    let nodes = builder.build(props);
+    return new TreeCollection(nodes, {expandedKeys: new Set()});
+  }, [builder, props]);
+
+  if (props.selectedKey && props.inputValue) {
+    let selectedItem = collection.getItem(props.selectedKey);
+    let itemText = selectedItem ? selectedItem.textValue : '';
+    if (itemText !== props.inputValue) {
+      console.error('Mismatch between selected item and inputValue!');
+    }
+  }
+
+  let lastSelectedKey = useRef('');
+  let onInputChange = (value) => {
+    if (props.onInputChange) {
+      props.onInputChange(value);
+    }
+
+    let newSelectedKey = computeKeyFromValue(value, collection);
+    if (newSelectedKey !== selectedKey) {
+      if (onSelectionChange) {
+        onSelectionChange(newSelectedKey);
+      }
+    }
+
+    lastSelectedKey.current = newSelectedKey;
+  };
+
+  let initialSelectedKeyText = collection.getItem(props.selectedKey)?.textValue;
+  let initialDefaultSelectedKeyText = collection.getItem(props.defaultSelectedKey)?.textValue;
+  let [inputValue, setInputValue] = useControlledState(toString(props.inputValue), initialSelectedKeyText || toString(props.defaultInputValue) || initialDefaultSelectedKeyText || '', onInputChange);
+
+  let selectedKey = props.selectedKey || computeKeyFromValue(inputValue, collection);
+  let selectedKeys = useMemo(() => selectedKey != null ? [selectedKey] : [], [selectedKey]);
+
+  let triggerState = useMenuTriggerState(props);
+  // Fires when user hits Enter or clicks
+  let setSelectedKey = useCallback((key) => {
```

Comment:
Porting comment from combobox followup pull:

Still uncertain about this setSelectedKey + the useEffect below for when prop.selectedKey is updated + its interactions with onInputChange.

Especially in controlled inputValue cases where changes to the selectedKey will call props.onInputChange

Reply:
As a whole, a bit iffy on this and the two useEffects below